### DISCUSSION TITLE
PostCSS: quais plugins usar?

### DISCUSSION THREAD

**Discussion by luisfmsouza** [frontendbr/forum#31] [Perguntas e Respostas]
Fala galera. Alguém ai usando **PostCSS**?

Quais plugins vocês tem utilizado nos projetos?

**Comment by felipefialho** [comment]
Opa, beleza @luisfmsouza?

Atualmente estou usando o [Rucksack](https://simplaio.github.io/rucksack) e o [nanoCSS](http://cssnano.co/).

Sobre o Rucksack, eu [escrevi um post](http://www.felipefialho.com/blog/2015/apresento-o-rucksack/) ano passado.

**Comment by afonsopacifer** [comment]
@luisfmsouza eu tenho usado o conjunto de plugins que vem no bolo doido do [cssnext](http://cssnext.io/)!

Eu meio que troquei os pre-processadores pelo PostCSS, isso para escrever CSS de forma mais nativa,
e nesse aspecto o pacote do [cssnext](http://cssnext.io/) cai como uma luva :)

**Comment by mksanti** [comment]
Eu gosto muito de trabalhar com .scss e quando vi o postcss tb gostei, então procurei um jeito de juntar os dois e to usando o precss + autoprefixer mais algun modulos do Rucksack.

**Comment by luisfmsouza** [comment]
Massa, eu não conhecia o Rucksack, vou dar uma fuçada lá!

Valeu pessoal!

**Comment by simaodeveloper** [comment]
@LFeh e o [autoprefixer](https://github.com/postcss/autoprefixer)

**Comment by felipefialho** [comment]
@simaodeveloper Não uso o Autoprefixer porque o Rucksack já vem com ele, hehe :p

**Comment by felipefialho** [comment]
Deixa eu chamar o deus do PostCSS aqui -> @fernandofleury

**Comment by fernandofleury** [comment]
Oi

**Comment by fernandofleury** [comment]
é uma pergunta muito subjetiva. qual seu objetivo com seu projeto? você quer só ter acesso as features de preprocessador com um tempo de compilação menor? então é https://github.com/jonathantneal/precss.

**Comment by luisfmsouza** [comment]
@fernandofleury meu foco é estudar e explorar o PostCSS, alguns plugins que o pessoal comentou eu encaixei em projetos pessoais meus.

Inclusive, alguém de vocês já usou o [postcss-font-magician](https://github.com/jonathantneal/postcss-font-magician) com Stylus?

**Comment by felquis** [comment]
Estou usando Less, depois uso apenas autoprefix.

**Comment by felipefialho** [comment]
@luisfmsouza Ainda não usei em projetos reais. Quem me mostrou essa ferramenta foi o @neysimoes.

**Comment by luisfmsouza** [comment]
@neysimoes para de jogar Xadrez e dá um luz aqui...

**Comment by fernandofleury** [comment]
cara, não tem muito segredo não. postCSS é JS, não tem estudo, eu pelo menos acho. o ideal é dar uma olhada nos plugins, e ver se tem algo que você se sente ais confortável pra escrever CSS, só isso.

Eu usei pra algumas coisas fora desse pipeline, tipo pra gerar uma [folha de estilos pra tipografia a partir dos glyphs](https://github.com/leroy-merlin-br/garden/blob/master/gulp/build-glyphs.js)

**Comment by fernandofleury** [comment]
como é JS, a limitação desse cara é o seu problema atual a ser resolvido, basicamente.

**Comment by simaodeveloper** [comment]
@LFeh aconteceu com você também, quando utiliza o [cssnano ](http://cssnano.co/)o mesmo remove o "charset 'utf-8'"?

**Comment by felipefialho** [comment]
@simaodeveloper Particularmente não uso o `charset 'utf-8'` no meu CSS, mas é possível configurar as opções do cssnano para ele gerar o CSS conforme você precisa.

**Comment by rickbenetti** [comment]
Caçando opções na web, encontrei o [PostStylus](https://github.com/seaneking/poststylus)

**Comment by felipefialho** [comment]
@rkb81 Estou usando o PostStylus para integrar o PostCSS no Stylus. Funciona muito bem.

**Comment by neysimoes** [comment]
To usando Stylus + PostCSS, mas sem usar o postStylus.
- Stylus
  - Rupture
- PostCSS
  - Lost
  - Rucksack (com autoprefixer)
- Plus
  - Gulp CSSnano

Primeiro rodo o Stylus com Rupture e depois vem o PostCSS em cima do CSS gerado.

PS: tá da mesma forma que o gh-pages do Lost grid
https://github.com/peterramsing/lost/blob/gh-pages/gulpfile.js

**Comment by felipefialho** [comment]
@neysimoes, carario mano, eu, você, o @Mendrone e o @willianjusten praticamente compartilha o Workflow. \o

**Comment by fernandofleury** [comment]
Me sinto mais confortável com sintaxe normal, principalmente por isso possibilitar o pipeline com folhas de estilo padrão também. O Stylus possibilita importar `.css` normal? Usei por um bom tempo, mas nunca tentei isso.

**Comment by luisfmsouza** [comment]
@neysimoes vc usa o font magician, ou encontrou alguma alternativa?

**Comment by felipefialho** [comment]
@fernandofleury 

Usando o @import ele renderiza como @import do CSS mesmo =(

Parece que é possível adicionar através das options do Stylus, mas nunca tentei.

**Comment by bernardodiasc** [comment]
Comecei a usar o CSS Modules, é uma sacada bem boa pra lidar com o (potencial) problema de cascata.

Resumindo a ideia, voce importa o arquivo css no template (js) e então usa as classes css como objetos js, cada nome de classe é transformado em outro nome único. Simples e prático, tão simples que da pra usar a mesma teoria direto no css pegando a treta de criar nome de classes únicos manualmente. Aliás é exatamente isso que muitas metodologias como BEM, SMACSS e SUIT tentam resolver, regras de nomenclatura pra manter a sanidade.

Tem uma apresentação de um dos criadores do bootstrap que passa correndo pela história do css desde 1994 até o css modules com ênfase no efeito cascata do css, super recomendo https://www.youtube.com/watch?v=iniwPUEbPUM

**Comment by rubenmarcus** [comment]
para projetos novos e escalaveis vocês recomendariam já iniciar com PostCSS ou SASS?

**Comment by bernardodiasc** [comment]
Bem legal essa lista http://postcss.parts/

**Comment by luisfmsouza** [comment]
@rubenmarcus se o projeto já vai nascer pra ser longo, vale a pena mesclar PostCSS e Preprocessador(Sass ou Stylus).

Autoprefixer e cssnano funcionariam bem nesse caso.

**Comment by rubenmarcus** [comment]
@luisfmsouza  Valeu! e pra quem está acostumado a fazer tudo em css puro, e já fez poucas coisas em SASS, existe algum tutorial recomendado, a curva de aprendizado é longa??

Se você tem um prazo curto tipo 1 mes pra fazer 10 telas por exemplo...
E não tem dominio de pré processadores
você
-CSS Puro
-SASS
-postcss + stylus?

**Comment by luisfmsouza** [comment]
Fala @rubenmarcus !

A curva de aprendizado de Sass e/ou Stylus é muito pequena, basicamente é só dar uma olhada na documentação pra ter em mente o que você pode fazer com eles, o principal é:
- variável
- nested
- mixins

No máximo, procura algum _get started_ no google. Mas acredite, é muito fácil! É praticamente CSS.

Se você tiver domínio de Gulp/Grunt ou até mesmo desenvoltura pra usar linha de comando, eu recomendaria somente o PostCSS, com [Rucksack](https://simplaio.github.io/rucksack/) ou [cssnext](http://cssnext.io/).

Utilizar Stylus + PostCSS também é uma boa ideia, assim você se acostuma com os dois de uma vez. Dá uma olhada nesse tutorial de instalação: http://webdesign.tutsplus.com/tutorials/using-postcss-together-with-sass-stylus-or-less--cms-24591

Qualquer coisa chama!

**Comment by rubenmarcus** [comment]
@luisfmsouza e talvez @fernandofleury   uma coisa aqui..

estou querendo desenvolver um projeto com esses plugins de Post-CSS:
Css Next: CSS 4 variaveis, Resposive Grid com media query range
Rucksack: Tipografia Responsiva , Property Aliases, Placeholder Styling,Font Src,Automatic Prefixing
Css Nano: Optimização
Style Lint
Svg Fallback
Responsive Images

Nao vou usar o Pre CSS , porque isso tudo ja me atende, ao que eu preciso, a minha dúvida é ,pra que usar o Stylus, seria dois passos a se fazer e eu ia perder todo o poder desses plugins se utilizasse o Stylus..

uma duvida eu to vendo os plugins de import mas eles só importam arquivos css.. eu queria poder separar os arquivos em variaveis e funcionalidades, e depois importar igual é no SASS, e tambem no final compilar um arquivo final só minificado, mas até agora não achei como.

**Comment by fernandofleury** [comment]
não entendi seu problema. porque não pode fazer `variables.css`?

**Comment by rubenmarcus** [comment]
@fernandofleury  
tem que usar um plugin de import certo?
https://github.com/postcss/postcss-import
https://github.com/jonathantneal/postcss-partial-import

qual o melhor dos dois?

hmm entendi, foi uma duvida bem leiga mesmo. Eu pensei que no arquivo final ele ia acabar gerando código sujo.
acho que o ultimo post foi por desencargo de consciencia porque to colocando  o hands on agora , pra um projeto grande, e quero fazer da melhor forma possivel

o postcss vcs usam algum plugin pra concatenação e minificação?

alias essa pratica de concatenar e minificar é correta certo?

**Comment by fernandofleury** [comment]
usa o `postcss-import`;

mas pra ser honesto, vai lá e faz um pouco do projeto... você tá se preocupando demais, e criando uma base enorme antes de se sentir confortável com ela.

se não quiser já no projeto principal, faz um projeto básico pra entender como funciona, senão isso não sai do papel.

só usar o cssnano que ele se responsabiliza pela minificação/otimização.

**Comment by luisfmsouza** [comment]
Boa Fleury! Coloca a mão na massa, qq coisa vai falando...

On Wednesday, 2 March 2016, Fleury notifications@github.com wrote:

> usa o postcss-import;
> 
> mas pra ser honesto, vai lá e faz um pouco do projeto... você tá se
> preocupando demais, e criando uma base enorme antes de se sentir
> confortável com ela.
> 
> se não quiser já no projeto principal, faz um projeto básico pra entender
> como funciona, senão isso não sai do papel.
> 
> só usar o cssnano que ele se responsabiliza pela minificação/otimização.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/frontendbr/forum/issues/31#issuecomment-191427988.

## 

Att,

_Luís Felipe Souza_
